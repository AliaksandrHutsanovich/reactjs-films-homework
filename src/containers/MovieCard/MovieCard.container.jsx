import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { REQUEST_TYPES } from '../../constants';
import { MovieCard } from '../../patterns';
import {
  useItemsSelector,
} from '../../hooks';

const MovieCardContainer = ({
  title,
  genresIds,
  rating,
  info,
  imgUrl,
  hoverable,
  renderHoverableContent,
  renderCoveringContent,
  id,
}) => {
  const [isShown, showCover] = useState(false);

  const genresById = useItemsSelector(REQUEST_TYPES.GENRES);

  return Object.keys(genresById).length ? (
    <MovieCard
      title={title}
      genres={genresIds.map((itemId) => (genresById[itemId].name))}
      rating={rating}
      info={info}
      imgUrl={imgUrl}
      hoverable={hoverable}
      renderHoverableContent={renderHoverableContent}
      renderCoveringContent={renderCoveringContent}
      showCoveringContent={isShown}
      onShowCover={showCover}
      id={id}
    />
  ) : null;
};

MovieCardContainer.propTypes = {
  title: PropTypes.string.isRequired,
  genresIds: PropTypes.arrayOf(
    PropTypes.number,
  ).isRequired,
  rating: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  imgUrl: PropTypes.string.isRequired,
  hoverable: PropTypes.bool,
  renderHoverableContent: PropTypes.func,
  renderCoveringContent: PropTypes.func,
  info: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
};

const renderDefaultContent = () => (<div />);

MovieCardContainer.defaultProps = {
  hoverable: false,
  renderHoverableContent: renderDefaultContent,
  renderCoveringContent: renderDefaultContent,
};

export default MovieCardContainer;
