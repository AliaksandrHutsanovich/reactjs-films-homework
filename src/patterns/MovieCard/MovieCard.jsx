import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { Card } from 'antd';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line

import MovieDescription from '../MovieDescription';

import styles from './MovieCard.scss';

const { Meta } = Card;

const MovieCard = ({
  title,
  genres,
  rating,
  imgUrl,
}) => (
  <Card
    className={styles.card}
    cover={<img alt="cardImg" src={imgUrl} className={styles.image} />}
  >
    <Meta
      description={(
        <MovieDescription
          title={title}
          genres={genres}
          rating={rating}
          ratingFontType="secondary_with_background"
        />
      )}
    />
  </Card>
);

MovieCard.propTypes = {
  title: PropTypes.string.isRequired,
  genres: PropTypes.arrayOf(
    PropTypes.string,
  ).isRequired,
  rating: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  imgUrl: PropTypes.string.isRequired,
};

export default memo(MovieCard); // additional memoization is possible
