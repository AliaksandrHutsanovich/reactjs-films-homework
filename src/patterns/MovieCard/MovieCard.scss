$cardWidth: 300;
$imageWidth: 450;

.card {
  width: #{$cardWidth}px;
  height: 320px;
  overflow: hidden;
}

.image {
  width: #{$imageWidth}px;
  margin: 0 -#{($imageWidth/2 - $cardWidth/2)}px;
}
