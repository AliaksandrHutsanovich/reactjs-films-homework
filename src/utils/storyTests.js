const storyTests = (stories, renderer) => {
  Object.keys(stories).forEach((storyKey) => {
    if (typeof stories[storyKey] !== 'function') {
      return;
    }
    it(`${storyKey} should be rendered properly`, () => {
      renderer.render(stories[storyKey]());
      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });
};

export default storyTests;
