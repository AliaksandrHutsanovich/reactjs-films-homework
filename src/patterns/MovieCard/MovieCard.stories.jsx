import React from 'react';
import MovieCard from './MovieCard';
import image from '../../assets/images/story-background.png';

export default {
  component: MovieCard,
  title: 'patterns/MovieCard',
};

const genres = [
  'Action',
  'Adventure',
  'Fantasy',
];

export const cardExample = () => (
  <MovieCard
    title="ASSASIN'S CREED"
    genres={genres}
    rating="4.7"
    imgUrl={image}
  />
);
