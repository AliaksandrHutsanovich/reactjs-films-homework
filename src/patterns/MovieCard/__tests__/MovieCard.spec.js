import ShallowRenderer from 'react-test-renderer/shallow';
import storyTests from '../../../utils/storyTests';
import * as stories from '../MovieCard.stories';

describe('<MovieCard />', () => {
  const renderer = new ShallowRenderer();

  storyTests(stories, renderer);
});
