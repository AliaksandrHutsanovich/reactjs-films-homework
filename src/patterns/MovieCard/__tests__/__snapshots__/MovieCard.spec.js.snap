// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard /> cardExample should be rendered properly 1`] = `
<Card
  className="card"
  cover={
    <img
      alt="cardImg"
      className="image"
      src="test-file-stub"
    />
  }
>
  <Meta
    description={
      <UNDEFINED
        genres={
          Array [
            "Action",
            "Adventure",
            "Fantasy",
          ]
        }
        rating="4.7"
        ratingFontType="secondary_with_background"
        title="ASSASIN'S CREED"
      />
    }
  />
</Card>
`;
